Audio Streaming

RoosterDragon opened the pull request:
Fixes #8593.

Music is now streamed! Changing music tracks no longer causes a lag spike or freezes the UI.

- Adds support for streaming in our audio engine via `Play2DStream`.
- Our sound manager now chooses to stream music, rather than caching it fully in memory (other sounds still use the caching mechanism).
- `AudFormat` is updated to support streaming, rather than having to load all the data into memory and expose a `MemoryStream`.

While testing, please listen out for:
- Choppy music (e.g. gaps in playback)
- Glitchy music (e.g. the music track seems to skip backwards a second or two and replay itself)
- Incorrect music (e.g. you try playing a music track and it plays another sound effect instead)

Other things to note:
- When you play a music track, it may take a few seconds to start. This is normal.
- If you switch music tracks several times really quickly, you will still be able to freeze the UI. This is a known limitation.

Finally, it would probably be worth also updating `VocFormat` and `WavFormat` in a similar way to support streaming audio - otherwise those formats won't benefit. Once my brain is less fried I may get to it.

RoosterDragon committed:
Add a streaming audio playback interface.

This allows audio to be streamed, rather than needed to be fully loaded into memory.

pchote commented:
I won't have time to look at this for a while, but two things to start with:
* travis-ci isn't happy
* My understanding is that AL will kill the source if its buffers empty before you queue more.  Won't that just kill the track instead of giving choppy playback?

RoosterDragon commented:
If the buffers run out, the source enters the stopped state. Queueing another buffer then starting the audio will get it going again.

The one catch is that playing it will restart from the beginning of any buffers in the source, therefore you need to dequeue the completed ones first to prevent them being repeated.

This is all handled starting [here](https://github.com/OpenRA/OpenRA/pull/13521/files#diff-40399e0400615b30cb4e205dc5cc1441R521).

reaperrr commented:
We definitely want this for next release, in my opinion.

pchote commented:
Ideally even the playtest, if time / review power allows.

reaperrr commented:
There is a slight 'hiccup' in playback after a few seconds, might be when transitioning from the initially loaded chunk to the main stream.

Other than that it seems to work fine, as far as I can tell.

pchote commented:
More often than not this will only play the first four seconds (two buffers?) of a track before it skips to the next, or repeats from the start if looping (e.g. the shellmap).  If I restart the game a few times the shellmap music will play properly, but the issue returns when I start a skirmish.

I suspect a race condition in the way you are scheduling buffer dequeuing, which seems to be confirmed by the problem going away if change the delay in the loop to `BufferSizeInSecs / 2`

RoosterDragon commented:
Looks like I forgot to account for the bit rate when creating the buffer. All the 16-bit audio therefore had buffers half the expected size.

This meant it would sleep for 2 seconds with only 2 seconds of audio buffered instead of the expected 4. This causes a slight hiccup as the audio runs out just as it starts queueing more of it. Fixing the buffer to be created with the correct byte size seems to fix that. I have changed the buffer size to one second to offset the memory increase.

Also fixed a bug with the reported seek position when audio streaming falls behind. If you change the delay from `TimeSpan.FromSeconds(BufferSizeInSecs)` to `TimeSpan.FromSeconds(BufferSizeInSecs * 4)` this will force 3 seconds of audio plus a 1 second gap (simulating slow audio streaming). Audio should simply pause briefly without skipping a track and without the time jumping around.

pchote commented:
> If you change the delay from TimeSpan.FromSeconds(BufferSizeInSecs) to TimeSpan.FromSeconds(BufferSizeInSecs * 4) this will force 3 seconds of audio plus a 1 second gap (simulating slow audio streaming). Audio should simply pause briefly without skipping a track and without the time jumping around.

For me this skips the track and the time jumps around.

pchote reviewed (commented):
Looks great to me otherwise, just a couple of minor nits.

I'll hold off my :+1: for now until we understand and hopefully have a solution for the track skips.

pchote commented:
Possibly related: On macOS we use the system OpenAL library instead of OpenAL Soft.

pchote commented:
It looks like the buffers being popped from `freeBuffers` are in the `AL_STOPPED` state, not `AL_INITIAL`.  The workaround to dequeue completed buffers then happily dequeues the newly queued buffers, and the cycle of queuing then immediately dequeuing repeats until we run out of song.

pchote commented on OpenRA.Platforms.Default/OpenAlSoundEngine.cs:
```diff
@@ -404,54 +412,256 @@ public void Dispose()
+									// If we resume playback from the stopped state, it resets to the beginning!
+									// To avoid replaying the same audio, we need to dequeue the processed buffers first.
+									if (state == AL10.AL_STOPPED)
+										DequeueBuffers(source, freeBuffers);
```
Commenting this out fixes my issues on OSX without introducing any new regressions as far as I can tell (it doesn't reset to the beginning or replay the same audio).

reaperrr commented:
> Finally, it would probably be worth also updating VocFormat and WavFormat in a similar way to support streaming audio - otherwise those formats won't benefit

I don't know about Voc (haven't checked), but for WavFormat, the IMA compression really needs this at some point, as decode time increases linearly with file size (which is bad for TDX' current soundtrack as well as RA2 music).
For uncompressed WAV I don't think it's necessary (it doesn't need to be decoded, so it loads near-instant even for large files, and using it for music probably isn't feasible anyway), but I have no idea if our WAV format loader is structured in a way that allows implementing this on a per-format-base.

RoosterDragon commented on OpenRA.Game/Sound/Sound.cs:
```diff
@@ -204,11 +206,17 @@ public void PlayMusicThen(MusicInfo m, Action then)
+
+			if (music == null)
+			{
+				onMusicComplete = null;
```
The current behaviour is to ignore the callback if music fails to load. I'm just taking the opportunity to also null out the callback since it will never get called. The callback is only invoked is music is playing and reaches completion.

The current uses are to switch to the next track automatically when the current one ends, and also as a callback from Lua (which I assume is for much the same purpose).

If we did want to change this, we'd want to make sure it didn't invoke infinite loops from failure to play things.

RoosterDragon commented on OpenRA.Platforms.Default/OpenAlSoundEngine.cs:
```diff
@@ -404,54 +412,256 @@ public void Dispose()
+						AL10.alBufferData(nextBuffer, format, data, new IntPtr(count), new IntPtr(sampleRate));
+						AL10.alSourceQueueBuffers(source, new IntPtr(1), ref nextBuffer);
+
+						lock (cts)
```
Added.

RoosterDragon commented on OpenRA.Platforms.Default/OpenAlSoundEngine.cs:
```diff
@@ -404,54 +412,256 @@ public void Dispose()
+			var offset = 0;
+			int count;
+			while (offset < buffer.Length &&
+				(count = await stream.ReadAsync(buffer, offset, buffer.Length - offset, cts.Token).ConfigureAwait(false)) > 0)
```
A sync read blocks the threadpool thread we are currently using until the read completes. An async read releases the thread to do other useful work (i.e. other tasks) whilst the stream is being read.

It's how async is supposed to work - free up threads that would otherwise just be blocked waiting for things and not otherwise performing useful computation.

RoosterDragon commented on OpenRA.Platforms.Default/OpenAlSoundEngine.cs:
```diff
@@ -404,54 +412,256 @@ public void Dispose()
+					int buffersProcessed;
+					AL10.alGetSourcei(Source, AL10.AL_BUFFERS_PROCESSED, out buffersProcessed);
+
+					return (totalBuffersPlayed + buffersProcessed) * BufferSizeInSecs;
```
The last buffer of the track might not be one full second long. Accounting for the actual number of samples in it fixes the accounting (I was doing this originally, should've commented it so I didn't forget why!)

reaperrr commented on OpenRA.Platforms.Default/OpenAlSoundEngine.cs:
```diff
@@ -404,54 +412,267 @@ public void Dispose()
+
+	class OpenAlStreamingSound : OpenAlSound
+	{
+		const int BufferCount = 3; 
```
> OpenRA.Platforms.Default/OpenAlSoundEngine.cs:L488: [SP2000] Invalid spacing at the end of the line.

Trailing space.

RoosterDragon committed:
Providing streaming AudFormat data.

AudFormat.GetPCMInputStream now returns data that is streamed, rather than a MemoryStream.

RoosterDragon commented:
I have a branch where I've changed `VocFormat` and `WavFormat` to stream too. But I need to give it some testing to make sure I didn't mess up (any handy files in the default mods that use these formats and actually last more than a second?). Will likely be a different PR at any rate.

Regarding the OSX issue: I'm at a lost to explain it. [The documentation](https://www.openal.org/documentation/) states that when a buffer is dequeued it transitions to an unused state. Requeueing the buffer should mark it as pending until it is payer through, as which points it becomes processed. This is the behaviour I am seeing on my system.

The current code to dequeue buffers in the stopped state is also correct. Try inserting a `await Task.Delay(TimeSpan.FromSeconds(BufferSizeInSecs * 2), cts.Token).ConfigureAwait(false);` after the call to queue buffers. When the sound resumes, it will jump back to the start of those buffers and replay audio you already heard. You would not see this behaviour by extending the delay before the dequeue buffers call, as once it completes it will immediately dequeue all the buffers. Thus the source will be empty and so it can't go backwards when you restart it.

Are you seeing this behaviour is all mods, or just d2k? (I can't test d2k music, not that I'm expecting the choice of mod to have an impact)

Bit annoying that it's environment specific, but we can let this PR sit around until there's time to work out why. Apart from the OSX peculiarity I believe I've addressed all the other problems listed.

abcdefg30 reviewed (approved):
Works ingame on my end in RA and D2k.

pchote commented:
> Try inserting a `await Task.Delay(TimeSpan.FromSeconds(BufferSizeInSecs * 2), cts.Token).ConfigureAwait(false);` after the call to queue buffers. When the sound resumes, it will jump back to the start of those buffers and replay audio you already heard.

When I try do this the sound never resumes.  Additional buffers appear to be queued (the play timer advances) but no more music is heard.

pchote commented:
Some probably related links:
* https://github.com/lunixbochs/aquaria/blob/master/BBGE/FmodOpenALBridge.cpp#L841
* http://www.openradar.me/8556020

Are we able to manually calculate/track whether we expect AL_INITIAL vs AL_STOPPED?

pchote commented:
IMO if something causes the audio thread to be starved for more than 3 seconds then we have much bigger problems than the music glitching on macOS.  That is a relatively benign bug without a simple fix, so I don't think we should block this feature over it.  If we *really* have to then I don't think it should be *that* hard to write a CoreAudio backend.

pchote commented on OpenRA.Game/Primitives/ReadOnlyAdapterStream.cs:
```diff
@@ -0,0 +1,78 @@
+﻿using System;
```
Missing file header here

pchote reviewed (approved):
LGTM aside from one issue.  I'll push a fix to your branch so we can merge this without waiting.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e1e4277b65bad9b5f4becaf6d69dd89516e5e003

Mailaender commented:
@RoosterDragon You can claim your bounty at https://www.bountysource.com/issues/23256122-sound-engine-should-support-streaming

RoosterDragon commented:
Best see if we need to revert it due to #13763 first I suppose.
